erpnext/public/js/controllers/transaction.js

```diff
@@ -903,20 +903,73 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 
 	shipping_rule: function() {
 		var me = this;
-		if(this.frm.doc.shipping_rule) {
+		if(this.frm.doc.shipping_rule) {			
+			if(localStorage) {
+				var last_shipping_rule = localStorage.getItem("last_shipping_rule");					
+				localStorage.removeItem("last_shipping_rule");
+			}			
 			return this.frm.call({
 				doc: this.frm.doc,
 				method: "apply_shipping_rule",
 				callback: function(r) {
 					if(!r.exc) {
+						if(last_shipping_rule && cur_frm.doc.shipping_rule != last_shipping_rule)	{													
+							me.remove_duplicate_taxes(last_shipping_rule,"shipping_rule"); 
+								if(localStorage) {
+									localStorage.setItem("last_shipping_rule", cur_frm.doc.shipping_rule);
+									}				
+						}
+						else{
+							if(localStorage) {
+								localStorage.setItem("last_shipping_rule", cur_frm.doc.shipping_rule);
+							}
+						}                           
 						me.calculate_taxes_and_totals();
 					}
 				}
 			}).fail(() => this.frm.set_value('shipping_rule', ''));
-		}
-		else {
-			me.calculate_taxes_and_totals();
-		}
+		}		
+	},
+
+	remove_duplicate_taxes: function(last_shipping_rule,field_type) {
+			return this.frm.call({
+				method: "erpnext.controllers.accounts_controller.get_shipping_rule_account_head",
+				args: {
+					"master_doctype": frappe.meta.get_docfield(this.frm.doc.doctype, "shipping_rule",
+						this.frm.doc.name).options,
+					"master_name": last_shipping_rule
+				},
+				callback: function(r) {	
+					if(!r.exc) {
+						if(r.message.account) {	
+							if(field_type == "shipping_rule")	
+							{			
+								var tbl = cur_frm.doc.taxes || [];
+								var i = tbl.length;
+									while (i--)
+										{
+											if(tbl[i].account_head == r.message.account)
+											{
+												cur_frm.get_field("taxes").grid.grid_rows[i].remove();
+												cur_frm.refresh_fields();
+											}
+										}
+							}else{													
+								var tbl = cur_frm.doc.taxes || [];						
+								var i = tbl.length;
+									while (i--)
+									{
+										if(tbl[i].account_head != r.message.account)
+										{
+											cur_frm.get_field("taxes").grid.grid_rows[i].remove();
+											cur_frm.refresh_fields();
+										}
+									}
+							}
+						}
+					}
+				}
+			});				
```

Comment:
- Hey as per our discussion, we were going to fetch the tax rules from the python function and then remove the duplicates. I cannot see that function